cmd/celvet/main.go

```diff
@@ -81,23 +91,55 @@ func main() {
 	}
 
 	limitErrors := celvet.CheckMaxLimits(structural)
+	costErrors, compileErrors, otherErrors := celvet.CheckExprCost(structural)
+	if useJSON {
+		emitJSON(limitErrors, costErrors, compileErrors, otherErrors)
+	} else {
+		emitText(limitErrors, costErrors, compileErrors, otherErrors, *humanReadable)
+	}
+
+	if len(limitErrors)+len(costErrors)+len(compileErrors) > 0 {
+		os.Exit(1)
+	}
+}
+
+func emitJSON(limitErrors []*celvet.LimitError, costErrors []*celvet.CostError, compileErrors []*celvet.CompilationError, otherErrors []error) {
+	type JSONOutput struct {
+		LimitErrors   []*celvet.LimitError       `json:"limitErrors,omitempty"`
+		CostErrors    []*celvet.CostError        `json:"costErrors,omitempty"`
+		CompileErrors []*celvet.CompilationError `json:"compileErrors,omitempty"`
+		OtherErrors   []string                   `json:"otherErrors,omitempty"`
+	}
+
+	encoder := json.NewEncoder(os.Stdout)
+	encoder.SetEscapeHTML(false)
+	strOtherErrors := make([]string, 0)
+	for _, err := range otherErrors {
+		strOtherErrors = append(strOtherErrors, err.Error())
+	}
+	output := JSONOutput{LimitErrors: limitErrors, CostErrors: costErrors, CompileErrors: compileErrors, OtherErrors: strOtherErrors}
+	err := encoder.Encode(output)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error generating JSON output: %s\n", err)
+		os.Exit(1)
```

Comment:
I wasn't expecting the function to exit. Consider only exiting in main and limiting output here to json. If there is an error is serializing json, return and error and let main handle it.

Reply:
Right, now `emitJSON` returns the error and only `main` exits - fixed